src/v/net/conn_quota.cc

```diff
@@ -15,11 +15,20 @@
 #include "rpc/logger.h"
 #include "seastar/core/coroutine.hh"
 #include "ssx/future-util.h"
+#include "vassert.h"
 #include "vlog.h"
 
+#include <seastar/core/smp.hh>
+
 namespace net {
 
 conn_quota::units::~units() noexcept {
+    vassert(
+      _owner_shard == ss::this_shard_id(),
+      "units must be created and destroyed on the same shard. Owner: {}, "
+      "destroyed on: {}",
+      _owner_shard,
+      ss::this_shard_id());
```

Comment:
we have a little utility that embodies this pattern: `src/v/bytes/oncore.h`

Reply:
Nice! Done.